finality-aleph/src/sync/task_queue.rs

```diff
@@ -0,0 +1,103 @@
+use std::{
+    cmp::Ordering,
+    collections::{binary_heap::PeekMut, BinaryHeap},
+    fmt::{Debug, Formatter},
+};
+
+use log::warn;
+use tokio::time::{sleep, Duration, Instant};
+
+#[derive(Clone, Eq, PartialEq)]
+struct ScheduledTask<T: Eq> {
+    task: T,
+    scheduled_time: Instant,
+}
+
+impl<T: Eq> PartialOrd for ScheduledTask<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T: Eq> Ord for ScheduledTask<T> {
+    /// Compare tasks so that earlier times come first in a max-heap.
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.scheduled_time.cmp(&self.scheduled_time)
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct TaskQueue<T: Eq + PartialEq> {
+    queue: BinaryHeap<ScheduledTask<T>>,
+}
+
+impl<T: Eq + PartialEq> Debug for TaskQueue<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TaskQueue")
+            .field("task count", &self.queue.len())
+            .finish()
+    }
+}
+
+/// Implements a queue allowing for scheduling tasks for some time in the future.
+impl<T: Eq> TaskQueue<T> {
+    /// Creates an empty queue.
+    pub fn new() -> Self {
+        Self {
+            queue: BinaryHeap::new(),
+        }
+    }
+
+    /// Schedules `task` for after `delay`.
+    pub fn schedule_in(&mut self, task: T, delay: Duration) {
+        let scheduled_time = match Instant::now().checked_add(delay) {
+            Some(time) => time,
+            None => {
+                warn!(target: "aleph-sync", "Could not schedule task in {:?}. Instant out of bound.", delay);
+                return;
+            }
+        };
+        self.queue.push(ScheduledTask {
+            task,
+            scheduled_time,
+        });
+    }
+
+    /// Awaits for the first and most overdue task and returns it. Returns `None` if there are no tasks.
+    pub async fn pop(&mut self) -> Option<T> {
+        let scheduled_task = self.queue.peek_mut()?;
+
+        sleep(
+            scheduled_task
+                .scheduled_time
+                .saturating_duration_since(Instant::now()),
+        )
+        .await;
```

Comment:
It might be worth it to avoid the sleep if the duration is zero, I'm not sure the implementation necessarily handles this efficiently. Although if it does it's probably better to leave as-is... <_<"